evidence/reactor_test.go

```diff
@@ -31,129 +33,172 @@ var (
 )
 
 type reactorTestSuite struct {
-	reactor *evidence.Reactor
-	pool    *evidence.Pool
-
-	peerID p2p.NodeID
-
-	evidenceChannel   *p2p.Channel
-	evidenceInCh      chan p2p.Envelope
-	evidenceOutCh     chan p2p.Envelope
-	evidencePeerErrCh chan p2p.PeerError
-
-	peerUpdatesCh chan p2p.PeerUpdate
-	peerUpdates   *p2p.PeerUpdates
+	network          *p2ptest.Network
+	logger           log.Logger
+	reactors         map[p2p.NodeID]*evidence.Reactor
+	pools            map[p2p.NodeID]*evidence.Pool
+	evidenceChannels map[p2p.NodeID]*p2p.Channel
+	peerUpdates      map[p2p.NodeID]*p2p.PeerUpdates
+	peerChans        map[p2p.NodeID]chan p2p.PeerUpdate
+	nodes            []*p2ptest.Node
+	numStateStores   int
 }
 
-func setup(t *testing.T, logger log.Logger, pool *evidence.Pool, chBuf uint) *reactorTestSuite {
+func setup(t *testing.T, stateStores []sm.Store, chBuf uint) *reactorTestSuite {
 	t.Helper()
 
 	pID := make([]byte, 16)
 	_, err := rng.Read(pID)
 	require.NoError(t, err)
 
-	peerUpdatesCh := make(chan p2p.PeerUpdate)
-
+	numStateStores := len(stateStores)
 	rts := &reactorTestSuite{
-		pool:              pool,
-		evidenceInCh:      make(chan p2p.Envelope, chBuf),
-		evidenceOutCh:     make(chan p2p.Envelope, chBuf),
-		evidencePeerErrCh: make(chan p2p.PeerError, chBuf),
-		peerUpdatesCh:     peerUpdatesCh,
-		peerUpdates:       p2p.NewPeerUpdates(peerUpdatesCh),
-		peerID:            p2p.NodeID(fmt.Sprintf("%x", pID)),
+		numStateStores: numStateStores,
+		logger:         log.TestingLogger().With("testCase", t.Name()),
+		network:        p2ptest.MakeNetwork(t, numStateStores),
+		reactors:       make(map[p2p.NodeID]*evidence.Reactor, numStateStores),
+		pools:          make(map[p2p.NodeID]*evidence.Pool, numStateStores),
+		peerUpdates:    make(map[p2p.NodeID]*p2p.PeerUpdates, numStateStores),
+		peerChans:      make(map[p2p.NodeID]chan p2p.PeerUpdate, numStateStores),
 	}
 
-	rts.evidenceChannel = p2p.NewChannel(
+	rts.evidenceChannels = rts.network.MakeChannelsNoCleanup(t,
 		evidence.EvidenceChannel,
 		new(tmproto.EvidenceList),
-		rts.evidenceInCh,
-		rts.evidenceOutCh,
-		rts.evidencePeerErrCh,
-	)
+		int(chBuf))
+	require.Len(t, rts.network.RandomNode().PeerManager.Peers(), 0)
 
-	rts.reactor = evidence.NewReactor(
-		logger,
-		rts.evidenceChannel,
-		rts.peerUpdates,
-		pool,
-	)
+	idx := 0
+	evidenceTime := time.Date(2019, 1, 1, 0, 0, 0, 0, time.UTC)
+	for nodeID := range rts.network.Nodes {
+		logger := rts.logger.With("validator", idx)
+		evidenceDB := dbm.NewMemDB()
+		blockStore := &mocks.BlockStore{}
+		blockStore.On("LoadBlockMeta", mock.AnythingOfType("int64")).Return(
+			&types.BlockMeta{Header: types.Header{Time: evidenceTime}},
+		)
```

Comment:
This should resolve the case with the lagging peer that seems to accept more evidence then it should be able to verify
```suggestion
		state, _ := stateStores[idx].Load()
		blockStore.On("LoadBlockMeta", mock.AnythingOfType("int64")).Return(func(h int64) *types.BlockMeta {
			if (h <= state.LastBlockHeight) {
				return &types.BlockMeta{Header: types.Header{Time: evidenceTime}}
			}	
			return nil
		})
```